doubleml/tests/test_nonlinear_score_mixin.py

```diff
@@ -0,0 +1,298 @@
+
+    def _check_score(self, score):
+        if isinstance(score, str):
+            valid_score = ['IV-type', 'partialling out', 'no_root_pos', 'no_root_neg']
+            if score not in valid_score:
+                raise ValueError('Invalid score ' + score + '. ' +
+                                 'Valid score ' + ' or '.join(valid_score) + '.')
+        else:
+            if not callable(score):
+                raise TypeError('score should be either a string or a callable. '
+                                '%r was passed.' % score)
+        return
+
+    def _check_data(self, obj_dml_data):
+        pass
+
+    def _nuisance_est(self, smpls, n_jobs_cv, return_models=False):
+        x, y = check_X_y(self._dml_data.x, self._dml_data.y,
+                         force_all_finite=False)
+        x, d = check_X_y(x, self._dml_data.d,
+                         force_all_finite=False)
+
+        # nuisance l
+        l_hat = _dml_cv_predict(self._learner['ml_l'], x, y, smpls=smpls, n_jobs=n_jobs_cv,
+                                est_params=self._get_params('ml_l'), method=self._predict_method['ml_l'])
+        _check_finite_predictions(l_hat['preds'], self._learner['ml_l'], 'ml_l', smpls)
+
+        # nuisance m
+        m_hat = _dml_cv_predict(self._learner['ml_m'], x, d, smpls=smpls, n_jobs=n_jobs_cv,
+                                est_params=self._get_params('ml_m'), method=self._predict_method['ml_m'])
+        _check_finite_predictions(m_hat['preds'], self._learner['ml_m'], 'ml_m', smpls)
+
+        # an estimate of g is obtained for the IV-type score and callable scores
+        g_hat = {'preds': None, 'models': None}
+        if 'ml_g' in self._learner:
+            # get an initial estimate for theta using the partialling out score
+            psi_a = -np.multiply(d - m_hat['preds'], d - m_hat['preds'])
+            psi_b = np.multiply(d - m_hat['preds'], y - l_hat['preds'])
+            theta_initial = -np.nanmean(psi_b) / np.nanmean(psi_a)
+            # nuisance g
+            g_hat = _dml_cv_predict(self._learner['ml_g'], x, y - theta_initial*d, smpls=smpls, n_jobs=n_jobs_cv,
+                                    est_params=self._get_params('ml_g'), method=self._predict_method['ml_g'])
+            _check_finite_predictions(g_hat['preds'], self._learner['ml_g'], 'ml_g', smpls)
+
+        psi_a, psi_b = self._score_elements(y, d, l_hat['preds'], m_hat['preds'], g_hat['preds'], smpls)
+        psi_elements = {'psi_a': psi_a,
+                        'psi_b': psi_b}
+        preds = {'predictions': {'ml_l': l_hat['preds'],
+                                 'ml_m': m_hat['preds'],
+                                 'ml_g': g_hat['preds']},
+                 'models': {'ml_l': l_hat['models'],
+                            'ml_m': m_hat['models'],
+                            'ml_g': g_hat['models']}}
+
+        return psi_elements, preds
+
+    def _score_elements(self, y, d, l_hat, m_hat, g_hat, smpls):
```

Comment:
Is there a reason the use `smpls` as input argument?

Reply:
It is currently not used. However, we wanted to make it available for computing the score elements to be able to support approaches where there is some nested cross-fitting. E.g., when we once implemented the second order orthogonal score of https://arxiv.org/abs/1711.00342, we needed `smpls` in this step. The code looked something like this
```python
        elif score == '2nd_order_dml':
            mult_v_est = np.zeros_like(d, dtype='float64')
            for train_index, test_index in smpls:
                second_v_est = np.mean(v_hat[train_index] ** 2)
                cube_v_est = np.mean(v_hat[train_index] ** 3) - 3 * np.mean(v_hat[train_index]) * np.mean(v_hat[train_index] ** 2)
                mult_v_est[test_index] = v_hat[test_index]**3 - 3 * second_v_est * v_hat[test_index] - cube_v_est

            score_a = -np.multiply(v_hat, mult_v_est)
        elif score == '2nd_order_dml_crossfit':
            mult_v_est = np.zeros_like(d, dtype='float64')
            for _, test_index in smpls:
                nested_smpls = KFold(n_splits=2, shuffle=True).split(np.zeros(len(test_index)))
                for nested_train_index, nested_test_index in nested_smpls:
                    this_train_ind = test_index[nested_train_index]
                    this_test_ind = test_index[nested_test_index]
                    second_v_est = np.mean(v_hat[this_train_ind] ** 2)
                    cube_v_est = np.mean(v_hat[this_train_ind] ** 3) - 3 * np.mean(v_hat[this_train_ind]) * np.mean(v_hat[this_train_ind] ** 2)
                    mult_v_est[this_test_ind] = v_hat[this_test_ind]**3 - 3 * second_v_est * v_hat[this_test_ind] - cube_v_est

            score_a = -np.multiply(v_hat, mult_v_est)
```

Reply:
Ah, thanks. Then it might be helpful to keep it.